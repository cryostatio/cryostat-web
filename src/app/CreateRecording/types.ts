/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { EventTemplate, KeyValue } from '@app/Shared/Services/api.types';
import { ValidatedOptions } from '@patternfly/react-core';

export type EventTemplateIdentifier = Pick<EventTemplate, 'name' | 'type'>;

interface _FormBaseData {
  name: string;
  template?: EventTemplateIdentifier;
  labels: KeyValue[];
  continuous: boolean;
  archiveOnStop: boolean;
  restart: boolean;
  duration: number;
  durationUnit: number;
  skipDurationCheck?: boolean;
  maxAge: number;
  maxAgeUnit: number;
  maxSize: number;
  maxSizeUnit: number;
  toDisk: boolean;
}

interface _FormValidationData {
  nameValid: ValidatedOptions;
  labelsValid: ValidatedOptions;
  durationValid: ValidatedOptions;
}

export type CustomRecordingFormData = _FormBaseData & _FormValidationData;
