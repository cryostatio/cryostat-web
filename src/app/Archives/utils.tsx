/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { RecordingDirectory, Target } from '@app/Shared/Services/api.types';

export const includesDirectory = (arr: RecordingDirectory[], dir: RecordingDirectory): boolean => {
  return arr.some((t) => t.connectUrl === dir.connectUrl);
};

export const indexOfDirectory = (arr: RecordingDirectory[], dir: RecordingDirectory): number => {
  let index = -1;
  arr.forEach((d, idx) => {
    if (d.connectUrl === dir.connectUrl) {
      index = idx;
    }
  });
  return index;
};

export const getTargetFromDirectory = (dir: RecordingDirectory): Target => {
  return {
    connectUrl: dir.connectUrl,
    alias: dir.jvmId,
    labels: [],
    annotations: {
      cryostat: [],
      platform: [],
    },
  };
};
