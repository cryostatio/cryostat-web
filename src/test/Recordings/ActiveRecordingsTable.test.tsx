/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import '@testing-library/jest-dom';
import { authFailMessage } from '@app/ErrorView/types';
import { DeleteActiveRecordings, DeleteOrDisableWarningType } from '@app/Modal/types';
import { ActiveRecordingsTable } from '@app/Recordings/ActiveRecordingsTable';
import {
  emptyActiveRecordingFilters,
  emptyArchivedRecordingFilters,
  TargetRecordingFilters,
} from '@app/Shared/Redux/Filters/RecordingFilterSlice';
import { RootState } from '@app/Shared/Redux/ReduxStore';
import { ActiveRecording, RecordingState, NotificationMessage, Target } from '@app/Shared/Services/api.types';
import { defaultServices, ServiceContext, Services } from '@app/Shared/Services/Services';
import { TargetService } from '@app/Shared/Services/Target.service';
import dayjs, { defaultDatetimeFormat } from '@i18n/datetime';
import { act, cleanup, screen, within } from '@testing-library/react';
import { of, Subject } from 'rxjs';
import { basePreloadedState, DEFAULT_DIMENSIONS, render, resize } from '../utils';

const mockConnectUrl = 'service:jmx:rmi://someUrl';
const mockJvmId = 'id';
const mockTarget = {
  connectUrl: mockConnectUrl,
  alias: 'fooTarget',
  jvmId: mockJvmId,
  labels: [],
  annotations: { cryostat: [], platform: [] },
};
const mockRecordingLabels = [
  {
    key: 'someLabel',
    value: 'someValue',
  },
];
const mockRecording: ActiveRecording = {
  name: 'someRecording',
  downloadUrl: 'http://downloadUrl',
  reportUrl: 'http://reportUrl',
  metadata: { labels: mockRecordingLabels },
  startTime: 1234567890,
  id: 0,
  state: RecordingState.RUNNING,
  duration: 1000, // 1000ms
  continuous: false,
  toDisk: false,
  maxSize: 0,
  maxAge: 0,
};
const mockAnotherRecording = { ...mockRecording, name: 'anotherRecording', id: 1 };
const mockCreateNotification = {
  message: { target: mockConnectUrl, recording: mockAnotherRecording, jvmId: mockJvmId },
} as NotificationMessage;
const mockLabelsNotification = {
  message: {
    target: mockConnectUrl,
    recordingName: 'someRecording',
    jvmId: mockJvmId,
    metadata: { labels: { someLabel: 'someUpdatedValue' } },
  },
} as NotificationMessage;
const mockStopNotification = {
  message: { target: mockConnectUrl, recording: mockRecording, jvmId: mockJvmId },
} as NotificationMessage;
const mockDeleteNotification = mockStopNotification;

jest.mock('@app/Recordings/RecordingFilters', () => {
  // Already tested separately
  return {
    ...jest.requireActual('@app/Recordings/RecordingFilters'),
    RecordingFilters: jest.fn(() => {
      return <div>RecordingFilters</div>;
    }),
  };
});

jest.spyOn(defaultServices.api, 'archiveRecording').mockReturnValue(of(true));
jest.spyOn(defaultServices.api, 'deleteRecording').mockReturnValue(of(true));
jest.spyOn(defaultServices.api, 'doGet').mockReturnValue(of([mockRecording]));
jest.spyOn(defaultServices.api, 'downloadRecording').mockReturnValue(void 0);
jest.spyOn(defaultServices.api, 'grafanaDashboardUrl').mockReturnValue(of('/grafanaUrl'));
jest.spyOn(defaultServices.api, 'grafanaDatasourceUrl').mockReturnValue(of('/datasource'));
jest.spyOn(defaultServices.api, 'stopRecording').mockReturnValue(of(true));
jest.spyOn(defaultServices.api, 'uploadActiveRecordingToGrafana').mockReturnValue(of(true));
jest.spyOn(defaultServices.target, 'target').mockReturnValue(of(mockTarget));
jest.spyOn(defaultServices.target, 'authFailure').mockReturnValue(of());

jest.spyOn(defaultServices.reports, 'delete').mockReturnValue(void 0);

jest
  .spyOn(defaultServices.settings, 'deletionDialogsEnabledFor')
  .mockReturnValueOnce(true) // shows a popup when Delete is clicked and then deletes the recording after clicking confirmation Delete
  .mockReturnValueOnce(false) // deletes the recording when Delete is clicked w/o popup warning
  .mockReturnValue(true);

jest.spyOn(defaultServices.settings, 'datetimeFormat').mockReturnValue(of(defaultDatetimeFormat));

jest
  .spyOn(defaultServices.notificationChannel, 'messages')
  .mockReturnValueOnce(of()) // renders the recording table correctly
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())

  .mockReturnValueOnce(of(mockCreateNotification)) // adds a recording table after receiving a notification
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())

  .mockReturnValueOnce(of()) // updates the recording labels after receiving a notification
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of(mockLabelsNotification))

  .mockReturnValueOnce(of()) // stops a recording after receiving a notification
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of(mockStopNotification))
  .mockReturnValueOnce(of())

  .mockReturnValueOnce(of()) // removes a recording after receiving a notification
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of(mockDeleteNotification))
  .mockReturnValue(of()); // all other tests

jest.spyOn(window, 'open').mockReturnValue(null);

describe('<ActiveRecordingsTable />', () => {
  let preloadedState: RootState;

  beforeAll(async () => {
    await act(async () => {
      resize(2400, 1080);
    });
  });

  beforeEach(async () => {
    mockRecording.metadata.labels = mockRecordingLabels;
    mockRecording.state = RecordingState.RUNNING;
    preloadedState = {
      ...basePreloadedState,
      recordingFilters: {
        list: [
          {
            target: mockTarget.connectUrl,
            active: {
              selectedCategory: 'Labels',
              filters: emptyActiveRecordingFilters,
            },
            archived: {
              selectedCategory: 'Name',
              filters: emptyArchivedRecordingFilters,
            },
          } as TargetRecordingFilters,
        ],
        _version: '0',
      },
    };
  });

  afterAll(() => {
    resize(DEFAULT_DIMENSIONS[0], DEFAULT_DIMENSIONS[1]);
  });

  afterEach(cleanup);

  it('renders the recording table correctly', async () => {
    const { user } = render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <ActiveRecordingsTable archiveEnabled={true} toolbarBreakReference={document.body} />,
          },
        ],
      },
      preloadedState: preloadedState,
    });

    ['Create', 'Edit Labels', 'Stop', 'Delete'].map((text) => {
      const button = screen.getByText(text);
      expect(button).toBeInTheDocument();
      expect(button).toBeVisible();
    });

    ['Name', 'Start Time', 'Duration', 'State', 'Labels'].map((text) => {
      const header = screen.getByText(text);
      expect(header).toBeInTheDocument();
      expect(header).toBeVisible();
    });

    const checkboxes = screen.getAllByRole('checkbox');
    expect(checkboxes.length).toBe(2);
    checkboxes.forEach((checkbox) => {
      expect(checkbox).toBeInTheDocument();
      expect(checkbox).toBeVisible();
    });

    const name = screen.getByText(mockRecording.name);
    expect(name).toBeInTheDocument();
    expect(name).toBeVisible();

    const startTime = screen.getByText(dayjs(mockRecording.startTime).tz('UTC').format('L LTS z'));
    expect(startTime).toBeInTheDocument();
    expect(startTime).toBeVisible();

    await act(async () => {
      await user.click(startTime);
    });

    const toolTip = await screen.findByText(dayjs(mockRecording.startTime).toISOString());
    expect(toolTip).toBeInTheDocument();
    expect(toolTip).toBeVisible();

    const duration = screen.getByText(
      mockRecording.continuous || mockRecording.duration === 0 ? 'Continuous' : `${mockRecording.duration / 1000}s`,
    );
    expect(duration).toBeInTheDocument();
    expect(duration).toBeVisible();

    const state = screen.getByText(mockRecording.state);
    expect(state).toBeInTheDocument();
    expect(state).toBeVisible();

    Object.keys(mockRecordingLabels).forEach((key) => {
      const label = screen.getByText(`${key}: ${mockRecordingLabels[key]}`);
      expect(label).toBeInTheDocument();
      expect(label).toBeVisible();
    });

    const actionIcon = within(screen.getByLabelText(`${mockRecording.name}-actions`)).getByLabelText('Actions');
    expect(actionIcon).toBeInTheDocument();
    expect(actionIcon).toBeVisible();
  });

  it('adds a recording after receiving a notification', async () => {
    render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <ActiveRecordingsTable archiveEnabled={true} />,
          },
        ],
      },
      preloadedState: preloadedState,
    });

    expect(screen.getByText('someRecording')).toBeInTheDocument();
    expect(screen.getByText('anotherRecording')).toBeInTheDocument();
  });

  it('updates the recording labels after receiving a notification', async () => {
    render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <ActiveRecordingsTable archiveEnabled={true} />,
          },
        ],
      },
      preloadedState: preloadedState,
    });

    expect(screen.getByText('someLabel: someUpdatedValue')).toBeInTheDocument();
    expect(screen.queryByText('someLabel: someValue')).not.toBeInTheDocument();
  });

  it('stops a recording after receiving a notification', async () => {
    render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <ActiveRecordingsTable archiveEnabled={true} />,
          },
        ],
      },
      preloadedState: preloadedState,
    });

    expect(screen.getByText('STOPPED')).toBeInTheDocument();
    expect(screen.queryByText('RUNNING')).not.toBeInTheDocument();
  });

  it('removes a recording after receiving a notification', async () => {
    render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <ActiveRecordingsTable archiveEnabled={true} />,
          },
        ],
      },
      preloadedState: preloadedState,
    });

    expect(screen.queryByText('someRecording')).not.toBeInTheDocument();
  });

  it('displays the toolbar buttons', async () => {
    render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <ActiveRecordingsTable archiveEnabled={true} toolbarBreakReference={document.body} />,
          },
        ],
      },
      preloadedState: preloadedState,
    });

    expect(screen.getByText('Create')).toBeInTheDocument();
    expect(screen.getByText('Archive')).toBeInTheDocument();
    expect(screen.getByText('Edit Labels')).toBeInTheDocument();
    expect(screen.getByText('Stop')).toBeInTheDocument();
    expect(screen.getByText('Delete')).toBeInTheDocument();
  });

  it('routes to the Create Flight Recording form when Create is clicked', async () => {
    const { user } = render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <ActiveRecordingsTable archiveEnabled={true} toolbarBreakReference={document.body} />,
          },
        ],
      },
      preloadedState: preloadedState,
    });

    await user.click(screen.getByText('Create'));

    // expect(history.entries.map((entry) => entry.pathname)).toStrictEqual(['/recordings', '/recordings/create']);
  });

  it('archives the selected recording when Archive is clicked', async () => {
    const { user } = render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <ActiveRecordingsTable archiveEnabled={true} toolbarBreakReference={document.body} />,
          },
        ],
      },
      preloadedState: preloadedState,
    });

    const checkboxes = screen.getAllByRole('checkbox');
    const selectAllCheck = checkboxes[0];
    await user.click(selectAllCheck);
    await user.click(screen.getByText('Archive'));

    const archiveRequestSpy = jest.spyOn(defaultServices.api, 'archiveRecording');

    expect(archiveRequestSpy).toHaveBeenCalledTimes(1);
    expect(archiveRequestSpy).toBeCalledWith('someRecording');
  });

  it('stops the selected recording when Stop is clicked', async () => {
    const { user } = render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <ActiveRecordingsTable archiveEnabled={true} toolbarBreakReference={document.body} />,
          },
        ],
      },
      preloadedState: preloadedState,
    });

    const checkboxes = screen.getAllByRole('checkbox');
    const selectAllCheck = checkboxes[0];
    await user.click(selectAllCheck);
    await user.click(screen.getByText('Stop'));

    const stopRequestSpy = jest.spyOn(defaultServices.api, 'stopRecording');

    expect(stopRequestSpy).toHaveBeenCalledTimes(1);
    expect(stopRequestSpy).toBeCalledWith('someRecording');
  });

  it('opens the labels drawer when Edit Labels is clicked', async () => {
    const { user } = render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <ActiveRecordingsTable archiveEnabled={true} toolbarBreakReference={document.body} />,
          },
        ],
      },
      preloadedState: preloadedState,
    });

    const checkboxes = screen.getAllByRole('checkbox');
    const selectAllCheck = checkboxes[0];
    await user.click(selectAllCheck);
    await user.click(screen.getByText('Edit Labels'));
    expect(screen.getByText('Edit Recording Labels')).toBeInTheDocument();
  });

  it('shows a popup when Delete is clicked and then deletes the recording after clicking confirmation Delete', async () => {
    const { user } = render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <ActiveRecordingsTable archiveEnabled={true} toolbarBreakReference={document.body} />,
          },
        ],
      },
      preloadedState: preloadedState,
    });

    const checkboxes = screen.getAllByRole('checkbox');
    const selectAllCheck = checkboxes[0];
    await user.click(selectAllCheck);

    await user.click(screen.getByText('Delete'));

    const deleteModal = await screen.findByLabelText(DeleteActiveRecordings.ariaLabel);
    expect(deleteModal).toBeInTheDocument();
    expect(deleteModal).toBeVisible();

    const deleteRequestSpy = jest.spyOn(defaultServices.api, 'deleteRecording');
    const dialogWarningSpy = jest.spyOn(defaultServices.settings, 'setDeletionDialogsEnabledFor');
    await user.click(screen.getByLabelText("Don't ask me again"));
    await user.click(within(screen.getByLabelText(DeleteActiveRecordings.ariaLabel)).getByText('Delete'));

    expect(deleteRequestSpy).toBeCalledTimes(1);
    expect(deleteRequestSpy).toBeCalledWith('someRecording');
    expect(dialogWarningSpy).toBeCalledTimes(1);
    expect(dialogWarningSpy).toBeCalledWith(DeleteOrDisableWarningType.DeleteActiveRecordings, false);
  });

  it('deletes the recording when Delete is clicked w/o popup warning', async () => {
    const { user } = render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <ActiveRecordingsTable archiveEnabled={true} toolbarBreakReference={document.body} />,
          },
        ],
      },
      preloadedState: preloadedState,
    });

    const checkboxes = screen.getAllByRole('checkbox');
    const selectAllCheck = checkboxes[0];
    await user.click(selectAllCheck);
    await user.click(screen.getByText('Delete'));

    const deleteRequestSpy = jest.spyOn(defaultServices.api, 'deleteRecording');

    expect(screen.queryByLabelText(DeleteActiveRecordings.ariaLabel)).not.toBeInTheDocument();
    expect(deleteRequestSpy).toHaveBeenCalledTimes(1);
    expect(deleteRequestSpy).toBeCalledWith('someRecording');
  });

  it('downloads a recording when Download Recording is clicked', async () => {
    const { user } = render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <ActiveRecordingsTable archiveEnabled={true} />,
          },
        ],
      },
      preloadedState: preloadedState,
    });

    await act(async () => {
      await user.click(within(screen.getByLabelText(`${mockRecording.name}-actions`)).getByLabelText('Actions'));
    });
    await user.click(screen.getByText('Download Recording'));

    const downloadRequestSpy = jest.spyOn(defaultServices.api, 'downloadRecording');

    expect(downloadRequestSpy).toHaveBeenCalledTimes(1);
    expect(downloadRequestSpy).toBeCalledWith(mockRecording);
  });

  it('uploads a recording to Grafana when View in Grafana is clicked', async () => {
    const { user } = render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <ActiveRecordingsTable archiveEnabled={true} />,
          },
        ],
      },
      preloadedState: preloadedState,
    });

    await act(async () => {
      await user.click(within(screen.getByLabelText(`${mockRecording.name}-actions`)).getByLabelText('Actions'));
    });
    await user.click(screen.getByText('View in Grafana ...'));

    const grafanaUploadSpy = jest.spyOn(defaultServices.api, 'uploadActiveRecordingToGrafana');

    expect(grafanaUploadSpy).toHaveBeenCalledTimes(1);
    expect(grafanaUploadSpy).toBeCalledWith('someRecording');
  });

  it('should show error view if failing to retrieve recordings', async () => {
    const subj = new Subject<void>();
    const mockTargetSvc = {
      target: () => of(mockTarget as Target),
      authFailure: () => subj.asObservable(),
    } as TargetService;
    const services: Services = {
      ...defaultServices,
      target: mockTargetSvc,
    };

    render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <ActiveRecordingsTable archiveEnabled={true} />,
          },
        ],
      },
      preloadedState: preloadedState,
      providers: [{ kind: ServiceContext.Provider, instance: services }],
    });

    await act(async () => subj.next());

    const failTitle = screen.getByText('Error retrieving recordings');
    expect(failTitle).toBeInTheDocument();
    expect(failTitle).toBeVisible();

    const authFailText = screen.getByText(authFailMessage);
    expect(authFailText).toBeInTheDocument();
    expect(authFailText).toBeVisible();

    const retryButton = screen.getByText('Retry');
    expect(retryButton).toBeInTheDocument();
    expect(retryButton).toBeVisible();

    const toolbar = screen.queryByLabelText('active-recording-toolbar');
    expect(toolbar).not.toBeInTheDocument();
  });
});
