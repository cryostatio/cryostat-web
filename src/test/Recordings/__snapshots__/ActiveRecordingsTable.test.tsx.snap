// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ActiveRecordingsTable /> renders correctly 1`] = `
<div
  className="pf-c-drawer pf-m-inline"
>
  <div
    className="pf-c-drawer__main"
  >
    <div
      className="pf-c-drawer__content recordings-table-drawer-content"
    >
      <div
        className="pf-c-drawer__body pf-m-padding"
      >
        <div
          className="pf-c-toolbar"
          data-ouia-component-id="OUIA-Generated-Toolbar-3"
          data-ouia-component-type="PF4/Toolbar"
          data-ouia-safe={true}
          id="active-recordings-toolbar"
        >
          <div
            className="pf-c-toolbar__content"
          >
            <div
              className="pf-c-toolbar__content-section"
            >
              <div>
                RecordingFilters
              </div>
              <div
                className="pf-c-toolbar__item"
              >
                <button
                  aria-disabled={false}
                  aria-label={null}
                  className="pf-c-button pf-m-primary"
                  data-ouia-component-id="OUIA-Generated-Button-primary-3"
                  data-ouia-component-type="PF4/Button"
                  data-ouia-safe={true}
                  disabled={false}
                  onClick={[Function]}
                  role={null}
                  type="button"
                >
                  Create
                </button>
              </div>
              <div
                className="pf-c-toolbar__item"
              >
                <button
                  aria-disabled={true}
                  aria-label={null}
                  className="pf-c-button pf-m-secondary pf-m-disabled"
                  data-ouia-component-id="OUIA-Generated-Button-secondary-5"
                  data-ouia-component-type="PF4/Button"
                  data-ouia-safe={true}
                  disabled={true}
                  onClick={[Function]}
                  role={null}
                  tabIndex={null}
                  type="button"
                >
                  Archive
                </button>
              </div>
              <div
                className="pf-c-toolbar__item"
              >
                <button
                  aria-disabled={true}
                  aria-label={null}
                  className="pf-c-button pf-m-secondary pf-m-disabled"
                  data-ouia-component-id="OUIA-Generated-Button-secondary-6"
                  data-ouia-component-type="PF4/Button"
                  data-ouia-safe={true}
                  disabled={true}
                  onClick={[Function]}
                  role={null}
                  tabIndex={null}
                  type="button"
                >
                  Edit Labels
                </button>
              </div>
              <div
                className="pf-c-toolbar__item"
              >
                <button
                  aria-disabled={true}
                  aria-label={null}
                  className="pf-c-button pf-m-tertiary pf-m-disabled"
                  data-ouia-component-id="OUIA-Generated-Button-tertiary-3"
                  data-ouia-component-type="PF4/Button"
                  data-ouia-safe={true}
                  disabled={true}
                  onClick={[Function]}
                  role={null}
                  tabIndex={null}
                  type="button"
                >
                  Stop
                </button>
              </div>
              <div
                className="pf-c-toolbar__item"
              >
                <button
                  aria-disabled={true}
                  aria-label={null}
                  className="pf-c-button pf-m-danger pf-m-disabled"
                  data-ouia-component-id="OUIA-Generated-Button-danger-3"
                  data-ouia-component-type="PF4/Button"
                  data-ouia-safe={true}
                  disabled={true}
                  onClick={[Function]}
                  role={null}
                  tabIndex={null}
                  type="button"
                >
                  Delete
                </button>
              </div>
            </div>
            <div
              className="pf-c-toolbar__expandable-content"
              id="active-recordings-toolbar-expandable-content-2"
            >
              <div
                className="pf-c-toolbar__group"
              />
            </div>
          </div>
          <div
            className="pf-c-toolbar__content pf-m-hidden"
            hidden={true}
          >
            <div
              className="pf-c-toolbar__group"
            />
          </div>
        </div>
        <table
          aria-label="Active Flight Recordings"
          className="pf-c-table pf-m-grid-md"
          data-ouia-component-id="OUIA-Generated-Table-1"
          data-ouia-component-type="PF4/Table"
          data-ouia-safe={true}
          role="grid"
        >
          <thead
            className=""
          >
            <tr
              className=""
              data-ouia-component-id="OUIA-Generated-TableRow-1"
              data-ouia-component-type="PF4/TableRow"
              data-ouia-safe={true}
              hidden={false}
            >
              <td
                className="pf-c-table__check"
                onMouseEnter={[Function]}
                scope={null}
              >
                <label>
                  <input
                    aria-label="Select all rows"
                    checked={false}
                    name="check-all"
                    onChange={[Function]}
                    type="checkbox"
                  />
                </label>
              </td>
              <th
                className=""
                onMouseEnter={[Function]}
                scope={null}
              />
              <th
                className=""
                onMouseEnter={[Function]}
                scope="col"
              >
                Name
              </th>
              <th
                className=""
                onMouseEnter={[Function]}
                scope="col"
              >
                Start Time
              </th>
              <th
                className=""
                onMouseEnter={[Function]}
                scope="col"
              >
                Duration
              </th>
              <th
                className=""
                onMouseEnter={[Function]}
                scope="col"
              >
                State
              </th>
              <th
                className=""
                onMouseEnter={[Function]}
                scope="col"
              >
                Labels
              </th>
              <th
                className=""
                onMouseEnter={[Function]}
                scope={null}
              />
            </tr>
          </thead>
          <tbody
            className=""
            role="rowgroup"
          >
            <tr
              className=""
              data-ouia-component-id="OUIA-Generated-TableRow-2"
              data-ouia-component-type="PF4/TableRow"
              data-ouia-safe={true}
              hidden={false}
            >
              <td
                className=""
              >
                <div
                  className="pf-c-check pf-m-standalone"
                >
                  <input
                    aria-invalid={false}
                    checked={false}
                    className="pf-c-check__input"
                    data-ouia-component-id="OUIA-Generated-Checkbox-1"
                    data-ouia-component-type="PF4/Checkbox"
                    data-ouia-safe={true}
                    disabled={false}
                    id="active-table-row-0-check"
                    name="active-table-row-0-check"
                    onChange={[Function]}
                    type="checkbox"
                  />
                </div>
              </td>
              <td
                aria-controls="active-ex-expand-0"
                className="pf-c-table__toggle"
                id="active-ex-toggle-0"
              >
                <button
                  aria-disabled={false}
                  aria-expanded={false}
                  aria-label="Details"
                  aria-labelledby="simple-node0 expand-toggle0"
                  className="pf-c-button pf-m-plain"
                  data-ouia-component-id="OUIA-Generated-Button-plain-1"
                  data-ouia-component-type="PF4/Button"
                  data-ouia-safe={true}
                  disabled={false}
                  id="expand-toggle0"
                  onClick={[Function]}
                  role={null}
                  type="button"
                >
                  <div
                    className="pf-c-table__toggle-icon"
                  >
                    <svg
                      aria-hidden={true}
                      aria-labelledby={null}
                      fill="currentColor"
                      height="1em"
                      role="img"
                      style={
                        Object {
                          "verticalAlign": "-0.125em",
                        }
                      }
                      viewBox="0 0 320 512"
                      width="1em"
                    >
                      <path
                        d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
                      />
                    </svg>
                  </div>
                </button>
              </td>
              <td
                className=""
                data-label="Name"
              >
                someRecording
              </td>
              <td
                className=""
                data-label="Start Time"
              >
                <span>
                  1970-01-15T06:56:07.890Z
                </span>
              </td>
              <td
                className=""
                data-label="Duration"
              >
                <span>
                  Continuous
                </span>
              </td>
              <td
                className=""
                data-label="State"
              >
                RUNNING
              </td>
              <td
                className=""
                data-label="Labels"
              >
                <span
                  className="pf-c-label"
                >
                  <span
                    className="pf-c-label__content"
                  >
                    someLabel: someValue
                  </span>
                </span>
              </td>
              <td
                className="pf-c-table__action"
                style={
                  Object {
                    "paddingRight": 0,
                  }
                }
              >
                <div
                  className="pf-c-dropdown pf-m-align-right"
                  data-ouia-component-id="OUIA-Generated-Dropdown-1"
                  data-ouia-component-type="PF4/Dropdown"
                  data-ouia-safe={true}
                >
                  <button
                    aria-expanded={false}
                    aria-haspopup={true}
                    aria-label="Actions"
                    className="pf-c-dropdown__toggle pf-m-plain"
                    disabled={false}
                    id="pf-dropdown-toggle-id-1"
                    onClick={[Function]}
                    onKeyDown={[Function]}
                    type="button"
                  >
                    <svg
                      aria-hidden={true}
                      aria-labelledby={null}
                      fill="currentColor"
                      height="1em"
                      role="img"
                      style={
                        Object {
                          "verticalAlign": "-0.125em",
                        }
                      }
                      viewBox="0 0 192 512"
                      width="1em"
                    >
                      <path
                        d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
                      />
                    </svg>
                  </button>
                </div>
              </td>
            </tr>
            <tr
              className="pf-c-table__expandable-row"
              data-ouia-component-id="OUIA-Generated-TableRow-3"
              data-ouia-component-type="PF4/TableRow"
              data-ouia-safe={true}
              hidden={true}
            >
              <td
                className=""
                colSpan={8}
                data-label="Content Details"
              >
                <div
                  className="pf-c-table__expandable-row-content"
                >
                  <p
                    className=""
                    data-ouia-component-id="OUIA-Generated-Text-1"
                    data-ouia-component-type="PF4/Text"
                    data-ouia-safe={true}
                    data-pf-content={true}
                  >
                    Recording Options:
                  </p>
                  <p
                    className=""
                    data-ouia-component-id="OUIA-Generated-Text-2"
                    data-ouia-component-type="PF4/Text"
                    data-ouia-safe={true}
                    data-pf-content={true}
                  >
                    toDisk = 
                    false
                       maxAge = 
                    0
                    s   maxSize = 
                    0
                    B
                  </p>
                  <br />
                  <hr />
                  <br />
                  <p
                    className=""
                    data-ouia-component-id="OUIA-Generated-Text-3"
                    data-ouia-component-type="PF4/Text"
                    data-ouia-safe={true}
                    data-pf-content={true}
                  >
                    Automated Analysis:
                  </p>
                  <span
                    aria-label="Contents"
                    aria-valuetext="Loading..."
                    className="pf-c-spinner pf-m-xl"
                    role="progressbar"
                  >
                    <span
                      className="pf-c-spinner__clipper"
                    />
                    <span
                      className="pf-c-spinner__lead-ball"
                    />
                    <span
                      className="pf-c-spinner__tail-ball"
                    />
                  </span>
                  <iframe
                    height="640"
                    hidden={true}
                    onLoad={[Function]}
                    title="Automated Analysis"
                    width="100%"
                  />
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div
      className="pf-c-drawer__panel pf-m-resizable"
      hidden={true}
      id="pf-drawer-panel-0"
      onTransitionEnd={[Function]}
    />
  </div>
</div>
`;
